accepted/2021/platform-dependencies/platform-dependencies.md

```diff
@@ -0,0 +1,433 @@
+    public class Component
+    {
+        // Name of the component
+        public string Name { get; set; }
+
+        // Type of the component
+        public ComponentType Type { get; set; }
+
+        // Set of platform dependencies this component has
+        public IList<PlatformDependency> Dependencies { get; }
+    }
+
+    // Types of .NET components
+    public enum ComponentType
+    {
+        // A shared framework (e.g. Microsoft.NETCore.App, Microsoft.AspNetCore.App)
+        SharedFramework,
+
+        // A NuGet package (e.g. System.Drawing.Common)
+        NuGetPackage
+    }
+
+    // Description of a platform dependency
+    public class PlatformDependency
+    {
+        // Name of the dependency artifact
+        public string Name { get; set; }
+
+        // Type of the dependency
+        public DependencyType Type { get; set; }
+
+        // Minimum version of the dependency artifact
+        public string? MinimumVersion { get; set; }
+
+        // A value indicating in what manner the dependency is used
+        public string Usage { get; set; }
+
+        // Reference to the dependency from the nearest parent platform that this dependency overrides
+        public DependencyRef? Overrides { get; set; }
+    }
+
+    // Types of dependencies
+    public enum DependencyType
+    {
+        // An executable file
+        Executable,
+        
+        // A library file (e.g. Windows DLL, Linux shared object library)
+        Library,
+
+        // A Linux package contained in a repository
+        LinuxPackage
+    }
+
+    // A reference to a dependency
+    public class DependencyRef
+    {
+        // Name of the referenced dependency
+        public string Name { get; set; }
+
+        // Type of the referenced dependency
+        public DependencyType DependencyType { get; set; }
+    }
+}
+```
+
+#### Platform
+
+Each platform has its own way of identifying assets that .NET is dependent upon. For example, the Open SSL library is labeled as openssl-libs in Fedora and libssl in Debian.  For that reason, platform dependencies can't be described without identifying which platform they are for. The model handles this by describing platforms as a top-level concept to organize dependencies. Since many dependencies apply to multiple versions of a given platform, the ability to describe a versionless platform, as well as more specific versioned platforms if necessary, is possible. This is done by associating a platform with a rid. Rids are hierarchical allowing more specific platforms to be targeted which provides an ideal mechanism to be able to flexibly describe the scope of dependencies. In the same way that rids are hierarchical, so too are the platforms in this model. This allows for less repetition in the actual model content and enables platforms with more specificity to append or amend dependencies from their parent platform. Some examples of this:
```

Comment:
In practice, I have seen a couple of issues in how RIDs are used that doesn't quite match this. In particular, RIDs are not scalable:  https://github.com/NuGet/Home/issues/5862

Reply:
Are there other issues besides the scalability problem?

I don't see the lack of RID scalability being an issue here.  I still think RIDs are the appropriate identifier to use here because it conceptually fits.  I don't want to be inventing a new hierarchy system.  And if it has the same limitations as the NuGet domain, to me that's fine; at least it's consistent.  I don't see this as a problem that this design should solve when NuGet itself hasn't solved it.

Reply:
Fair enough. I wanted to call it out because we are actively trying to find a fix/solution for this.

I am not aware of any non-scalability problems.

Reply:
Ok, thanks for calling it out as I wasn't aware of it.  I'll keep my eyes on it so that any changes there could potentially be incorporated here.